modules/core/04-channel/keeper/upgrade.go

```diff
@@ -95,6 +96,188 @@ func (k Keeper) WriteUpgradeInitChannel(
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		if upgradeSequence != counterpartyUpgradeSequence {
+			errorReceipt := types.NewErrorReceipt(upgradeSequence, errorsmod.Wrapf(types.ErrInvalidUpgradeSequence, "upgrade sequence %d was not smaller than the counter party chain upgrade sequence %d", upgradeSequence, counterpartyUpgradeSequence))
+			upgradeSequence++
+
+			k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+			k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
+			return 0, "", errorsmod.Wrapf(types.ErrUpgradeAborted, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		// if there is a crossing hello, i.e an UpgradeInit has been called on both channelEnds,
+		// then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+		// except for the channel state (upgrade channel will be in TRYUPGRADE and current channel will be in INITUPGRADE)
+		// if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
+		channel.State = types.TRYUPGRADE
+		k.SetChannel(ctx, portID, channelID, channel)
+
+		if !reflect.DeepEqual(channel, proposedUpgradeChannel) {
+			// TODO: log and emit events
+			if err := k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrInvalidChannel); err != nil {
+				return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, err.Error())
+			}
+			return 0, "", errorsmod.Wrap(types.ErrUpgradeAborted, "proposed upgrade channel did not equal expected channel")
+		}
+
+		// retrieve restore channel to return previous version
+		restoreChannel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
+		if !found {
+			// TODO(this case should be unreachable): write error receipt for upgrade sequence and abort / cancel upgrade
+			return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "upgrade aborted, error receipt written for upgrade sequence: %d", upgradeSequence)
+		}
+
+		return upgradeSequence, restoreChannel.Version, nil
+	default:
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s] but got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	return upgradeSequence, channel.Version, nil
+}
+
+// WriteUpgradeTryChannel writes a channel which has successfully passed the UpgradeTry handshake step.
+// An event is emitted for the handshake step.
+func (k Keeper) WriteUpgradeTryChannel(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	upgradeSequence uint64,
+	channelUpgrade types.Channel,
```

Comment:
#3401 bug exists here as well. I know it isn't clean, but I actually think the channel end should be retrieved from state instead of writing the proposed upgrade channel. This comes from a general security principal that you should be recreating the data you want to write to disk rather than writing the user input (as in this case it is unsafe without an additional check). This is partially why we do this for [openAck](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/keeper/handshake.go#L275) and openConfirm. I'm open to differing opinions, but it would be nice to be defensive here, if another field is added to the channel end which cannot be modified in an upgrade, we would need to remember to add another defensive check

Reply:
What if we add `proposedChannelUpgradeArgs` or some similar type which contains all the fields that can be modified. This can be passed as an argument to `MsgChannelUpgradeInit`, `ChanUpgradeInit`, `OnChanUpgradeInit` and `WriteChannelUpgradeInit`. It seems redundant to check that the `proposedUpgradeChannel` hasn't modified certain fields, seems safer to simply exclude them from consideration by adding an extra type

Reply:
nice catch, but do we need to create a separate type for this? We can fetch the existing channel and then modify it based on values present in `channelUpgrade`. A `proposedChannelUpgradeInfo` would surely contain the same fields. I do think it's a good idea to not set the channel directly though.

Reply:
The benefit of a separate type I can think of is that not all fields are upgradable, and so some might end up being ignored. With a custom type we can provide only fields that are valid to upgrade.

Reply:
Yes exactly. The benefit of that is that there is less room for error if the submitter specifies the non-modifiable fields incorrectly. We could use a similar syntax to client state `ZeroCustomFields()` function which returns a client state with the custom fields zeroed out (except we would do the opposite). I'm happy to long as we aren't setting the `proposedUpgradeChannel` directly into state and are simply modifying the existing one

Reply:
I think a helper like that is the best option, this PR is getting a little unwieldy so I think I will create a follow up for this if you're okay with it!